File: Challenge-ViewLift/app/src/main/java/com/example/user/challenge_viewlift/data/RemoteDataSource.java
package com.example.user.challenge_viewlift.data;

import com.example.user.challenge_viewlift.model.Snagfilms;

import io.reactivex.Observable;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by user on 2/5/18.
 */

public class RemoteDataSource {

    public static final String BASE_URL = "http://www.snagfilms.com/";
    public static final String DEFAULT_LIMIT = "20";

    public static Retrofit create() {

        return new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();
    }

    public static Observable<Snagfilms> getResponse(String limit) {
        RemoteService remoteService = create().create(RemoteService.class);
        return remoteService.getFilms(limit);
    }
}

File: WikiChallenge/app/src/main/java/com/example/user/myapplication/RemoteDataSource.kt
package com.example.user.myapplication

import io.reactivex.Observable
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.scalars.ScalarsConverterFactory
import retrofit2.http.GET
import retrofit2.http.Query

class RemoteDataSource {

    val baseURl = "https://en.wikipedia.org/"
    val action = "query"
    val list = "search"
    var srsearch: String? = "Nelson Mendela"
    val format = "json"


    private fun createRetrofit(): Retrofit {
        return Retrofit.Builder()
            .baseUrl(baseURl)
            //.addConverterFactory(GsonConverterFactory.create())
            .addConverterFactory(ScalarsConverterFactory.create())
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .build()

    }

    private fun getService(): RemoteService {
        return createRetrofit().create(RemoteService::class.java)
    }

    fun getResponse(srsearch: String): Observable<String> {
        return getService().getResponse(action, list, srsearch, format)
    }

    interface RemoteService {

        @GET("w/api.php")
        fun getResponse(
            @Query("action") action: String,
            @Query("list") list: String,
            @Query("srsearch") srsearch: String,
            @Query("format") format: String
        ): Observable<String>
    }
}

